awscli/autoprompt/logger.py

```diff
@@ -0,0 +1,64 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import logging
+from awscli.logger import LOG_FORMAT
+
+from prompt_toolkit.document import Document
+from prompt_toolkit.application import get_app
+
+
+class PromptToolkitHandler(logging.StreamHandler):
+
+    def emit(self, record):
+        try:
+            app = get_app()
+            if app.is_running and getattr(app, 'debug', False):
+                msg = self.format(record)
+                debug_buffer = app.layout.get_buffer_by_name('debug_buffer')
+                current_document = debug_buffer.document.text
+                if current_document:
+                    msg = '\n'.join([current_document, msg])
+                debug_buffer.set_document(
+                    Document(text=msg), bypass_readonly=True
+                )
+            else:
+                super().emit(record)
+        except:
+            self.handleError(record)
+
+
+class LoggersHandlerSwitcher:
+    def __init__(self, handler=PromptToolkitHandler(),
+                 log_level=logging.DEBUG,
+                 format_string=LOG_FORMAT):
+        self._handler = handler
+        self._handler.setLevel(log_level)
+        self._handler.setFormatter(logging.Formatter(format_string))
+        self._old_handlers = {}
```

Comment:
So I'm not sure how reusable this class is because it looks like we do not clear the `old_handlers` cache when we exit the context manager. If we want to continue making this a generalized abstraction, I think we should clear this dictionary when we exit.